Add bits tensor types

angelayi opened the pull request:
Test Plan: CI

Differential Revision: D41638739

vadimkantorov commented:
Lots of discussion about this and of useful ops on BitTensor: https://github.com/pytorch/pytorch/issues/32867 @ezyang

angelayi committed:
Add bits tensor types (#89990)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/89990

Original PR: https://github.com/pytorch/pytorch/pull/88594

Test Plan: CI

Reviewed By: Chillee

Differential Revision: D41638739

fbshipit-source-id: 98f84de978a8e7a617e96123b50be911129c5f6b

vkuzo commented:
put up https://github.com/pytorch/pytorch/pull/94992 with copy-pasta of this + clang9 fix
